TBot/Program.cs

```diff
@@ -1864,97 +1847,111 @@ class Program {
-										.OrderBy(c => Helpers.CalcDistance(c.Coordinate, target.Celestial.Coordinate, serverData))
-										.OrderBy(planet => planet.Coordinate.Type == Celestials.Moon).ToList();
+							List<Celestial> tempCelestials = (settings.AutoFarm.Origin.Length > 0) ? Helpers.ParseCelestialsList(settings.AutoFarm.Origin, celestials) : celestials;
+							List<Celestial> closestCelestials = tempCelestials
+								.OrderByDescending(planet => planet.Coordinate.Type == Celestials.Moon)
+								.OrderBy(c => Helpers.CalcDistance(c.Coordinate, target.Celestial.Coordinate, serverData)).ToList();
+
+							Celestial fromCelestial = null;
+							foreach (var c in closestCelestials) {
+								var tempCelestial = UpdatePlanet(c, UpdateType.Ships);
+								if (tempCelestial.Ships != null && tempCelestial.Ships.GetAmount(cargoShip) >= (numCargo + settings.AutoFarm.MinCargosToKeep)) {
+									var distance = Helpers.CalcDistance(tempCelestial.Coordinate, target.Celestial.Coordinate, serverData);
+									if (Helpers.IsSettingSet(settings.AutoFarm.MaxFlightDistance) && distance > settings.AutoFarm.MaxFlightDistance) continue;
+									fromCelestial = tempCelestial;
+									break;
 								}
-							} else {
-								closestCelestials = celestials
-									.OrderBy(c => Helpers.CalcDistance(c.Coordinate, target.Celestial.Coordinate, serverData))
-									.OrderBy(planet => planet.Coordinate.Type == Celestials.Moon).ToList();
 							}
-							foreach (var closest in closestCelestials) {
-								Celestial tempCelestial = closest;
-								tempCelestial = UpdatePlanet(tempCelestial, UpdateType.Ships);
 
+							if (fromCelestial == null) {
 								// TODO Future: If prefered cargo ship is not available or not sufficient capacity, combine with other cargo type.
-								if (tempCelestial.Ships.GetAmount(cargoShip) < numCargo) {
-									if (tempCelestial.Ships.GetAmount(cargoShip) >= (long) settings.AutoFarm.MinCargosToSend) {
-										numCargo = Helpers.CalcShipNumberForPayload(loot, cargoShip, researches.HyperspaceTechnology, userInfo.Class, serverData.ProbeCargo);
-									} else {
-										Helpers.WriteLog(LogType.Info, LogSender.AutoFarm, $"Insufficient {cargoShip.ToString()} on {tempCelestial.Coordinate}, require {numCargo} {cargoShip.ToString()}.");
+								foreach (var closest in closestCelestials) {
+									Celestial tempCelestial = closest;
+									tempCelestial = UpdatePlanet(tempCelestial, UpdateType.Ships);
+
+									if (tempCelestial.Ships.GetAmount(cargoShip) < numCargo + (long) settings.AutoFarm.MinCargosToKeep
+										&& Helpers.CalcDistance(tempCelestial.Coordinate, target.Celestial.Coordinate, serverData) < settings.AutoFarm.MaxFlightDistance) {
+
+										if (Helpers.IsSettingSet(settings.AutoFarm.BuildCargos) && settings.AutoFarm.BuildCargos == true) {
+											var neededCargos = numCargo + (long) settings.AutoFarm.MinCargosToKeep - tempCelestial.Ships.GetAmount(cargoShip);
```

Comment:
- [ ] AutoFarm Exception: Object reference not set to an instance of an object. Check and fix.

Reply:
Probably tempCelestial.Ships which requires UpdateShips()